src/Microsoft.Extensions.ApiDescription.Client/DownloadFile.cs

```diff
@@ -94,19 +96,116 @@ public async Task<bool> ExecuteAsync()
+            httpClient.Timeout = TimeSpan.FromMinutes(1);
+
+            var destinationExists = File.Exists(destinationPath);
+            var reachedCopy = false;
+            try
+            {
+                using (var response = await httpClient.GetAsync(uri, cancellationToken))
+                {
+                    response.EnsureSuccessStatusCode();
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    using (var responseStreamTask = response.Content.ReadAsStreamAsync())
+                    {
+                        var finished = await Task.WhenAny(
+                            responseStreamTask,
+                            Task.Delay(TimeSpan.FromSeconds(timeoutSeconds)));
+
+                        if (!ReferenceEquals(responseStreamTask, finished))
+                        {
+                            throw new TimeoutException($"Download failed to complete in {timeoutSeconds} seconds.");
+                        }
+
+                        var responseStream = await responseStreamTask;
+                        if (destinationExists)
+                        {
+                            // Check hashes before using the downloaded information.
+                            var downloadHash = GetHash(responseStream);
+                            responseStream.Position = 0L;
+
+                            byte[] destinationHash;
+                            using (var destinationStream = File.OpenRead(destinationPath))
+                            {
+                                destinationHash = GetHash(destinationStream);
+                            }
+
+                            var sameHashes = downloadHash.Length == destinationHash.Length;
+                            for (var i = 0L; sameHashes && i < downloadHash.Length; i++)
+                            {
+                                sameHashes = downloadHash[i] == destinationHash[i];
+                            }
+
+                            if (sameHashes)
+                            {
+                                log.LogMessage($"Not overwriting existing and matching file '{destinationPath}'.");
+                                return;
+                            }
+                        }
+                        else
+                        {
+                            // May need to create directory to hold the file.
+                            var destinationDirectory = Path.GetDirectoryName(destinationPath);
+                            if (!(string.IsNullOrEmpty(destinationDirectory) || Directory.Exists(destinationDirectory)))
+                            {
+                                Directory.CreateDirectory(destinationDirectory);
+                            }
+                        }
+
+                        // Create or overwrite the destination file.
+                        reachedCopy = true;
+                        using (var outStream = File.Create(destinationPath))
```

Comment:
Should this be written to asynchronously since we're already in an async method?